### ISSUE TITLE
Fails on Windows if path to this lib contains non-ascii chars

### ISSUE THREAD

**Issue by ostrolucky** [amphp/parallel#40]
On fresh git clone && composer install:
```
Správca@GADELAT-PC C:\
$ php parallel-functions\examples\1-simple-function.php
Took 3.409206867218 milliseconds.

Správca@GADELAT-PC C:\
$ mv parallel-functions á

Správca@GADELAT-PC C:\
$ php á\examples\1-simple-function.php
PHP Fatal error:  Uncaught Amp\MultiReasonException: Multiple errors encountered in C:\á\src\functions.php:59
Stack trace:
#0 [internal function]: Amp\ParallelFunctions\{closure}()
#1 C:\á\vendor\amphp\amp\lib\Coroutine.php(74): Generator->send(Array)
#2 C:\á\vendor\amphp\amp\lib\Internal\Placeholder.php(127): Amp\Coroutine->Amp\{closure}(NULL, Array)
#3 C:\á\vendor\amphp\amp\lib\Deferred.php(41): class@anonymous->resolve(Array)
#4 C:\á\vendor\amphp\amp\lib\functions.php(415): Amp\Deferred->resolve(Array)
#5 C:\á\vendor\amphp\amp\lib\Internal\ResolutionQueue.php(51): Amp\Promise\{closure}(Object(Amp\Parallel\Context\ContextException), NULL)
#6 C:\á\vendor\amphp\amp\lib\Failure.php(26): Amp\Internal\ResolutionQueue->__invoke(Object(Amp\Parallel\Context\ContextException), NULL)
#7 C:\á\vendor\amphp\amp\lib\Internal\Placeholder.php(122): Amp\Failure->onResolve(Object(Amp\Internal\ResolutionQueue))
#8 C:\á\vendor\amphp\amp\lib\Internal\Placeholder.php(151): Amp\Coroutine->resolve(Object(Amp\Failure))
#9 C:\á\vendor\amphp in C:\á\src\functions.php on line 59

Fatal error: Uncaught Amp\MultiReasonException: Multiple errors encountered in C:\á\src\functions.php on line 59

Amp\MultiReasonException: Multiple errors encountered in C:\á\src\functions.php on line 59

Call Stack:
    0.2040     381376   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2550    1547464   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:14


Správca@GADELAT-PC C:\
```

I streamlined bug to this library thanks to different error message I receive in other project where I use AMP

```
In Process.php line 177:

  [Amp\Parallel\Context\ContextException]
  The context stopped responding, potentially due to a fatal error or calling exit


Exception trace:
 Amp\Parallel\Context\Process->Amp\Parallel\Context\{closure}() at n/a:n/a
 Generator->throw() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Coroutine.php:71
 Amp\Coroutine->Amp\{closure}() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Failure.php:26
 Amp\Failure->onResolve() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Internal\Placeholder.php:122
 Amp\Coroutine->resolve() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Internal\Placeholder.php:151
 Amp\Coroutine->fail() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Coroutine.php:93
 Amp\Coroutine->Amp\{closure}() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Internal\Placeholder.php:127
 class@anonymous\C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Deferred.php004402D3->resolve() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Deferred.php:41
 Amp\Deferred->resolve() at C:\Users\Správca\src\php\Project\vendor\amphp\byte-stream\lib\ResourceInputStream.php:72
 Amp\ByteStream\ResourceInputStream::Amp\ByteStream\{closure}() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\NativeDriver.php:172
 Amp\Loop\NativeDriver->selectStreams() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\NativeDriver.php:68
 Amp\Loop\NativeDriver->dispatch() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\Driver.php:130
 Amp\Loop\Driver->tick() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\Driver.php:70
 Amp\Loop\Driver->run() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop.php:76
 Amp\Loop::run() at C:\Users\Správca\src\php\Project\bin\project:53
 Closure->{closure}() at n/a:n/a
 call_user_func() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Command\Command.php:250
 Symfony\Component\Console\Command\Command->run() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:865
 Symfony\Component\Console\Application->doRunCommand() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:241
 Symfony\Component\Console\Application->doRun() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:143
 Symfony\Component\Console\Application->run() at C:\Users\Správca\src\php\Project\bin\project:59

In ChannelParser.php line 61:

  [Amp\Parallel\Sync\ChannelException]
  Invalid packet received: Could not open input file: C:\Users\Spr\xc4\x82\xcb\x87vca\src\php\Project\vendor\amphp\parallel\lib\Context/Internal/proces
  s-runner.php\xa


Exception trace:
 Amp\Parallel\Sync\ChannelParser::parser() at n/a:n/a
 Generator->send() at C:\Users\Správca\src\php\Project\vendor\amphp\parser\lib\Parser.php:102
 Amp\Parser\Parser->push() at C:\Users\Správca\src\php\Project\vendor\amphp\parallel\lib\Sync\ChannelledStream.php:71
 Amp\Parallel\Sync\ChannelledStream->Amp\Parallel\Sync\{closure}() at n/a:n/a
 Generator->send() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Coroutine.php:74
 Amp\Coroutine->Amp\{closure}() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Internal\Placeholder.php:127
 class@anonymous\C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Deferred.php004402D3->resolve() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Deferred.php:41
 Amp\Deferred->resolve() at C:\Users\Správca\src\php\Project\vendor\amphp\byte-stream\lib\ResourceInputStream.php:72
 Amp\ByteStream\ResourceInputStream::Amp\ByteStream\{closure}() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\NativeDriver.php:172
 Amp\Loop\NativeDriver->selectStreams() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\NativeDriver.php:68
 Amp\Loop\NativeDriver->dispatch() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\Driver.php:130
 Amp\Loop\Driver->tick() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop\Driver.php:70
 Amp\Loop\Driver->run() at C:\Users\Správca\src\php\Project\vendor\amphp\amp\lib\Loop.php:76
 Amp\Loop::run() at C:\Users\Správca\src\php\Project\bin\project:53
 Closure->{closure}() at n/a:n/a
 call_user_func() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Command\Command.php:250
 Symfony\Component\Console\Command\Command->run() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:865
 Symfony\Component\Console\Application->doRunCommand() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:241
 Symfony\Component\Console\Application->doRun() at C:\Users\Správca\Downloads\symfony\src\Symfony\Component\Console\Application.php:143
 Symfony\Component\Console\Application->run() at C:\Users\Správca\src\php\Project\bin\project:59
```

Tested on Windows 7

**Comment by kelunik** [comment]
Could you please catch the `Amp\MultiReasonException` exception and dump `$e->getReasons()`?

**Comment by ostrolucky** [comment]
```
C:\á\examples\1-simple-function.php:17:
array(3) {
  [0] =>
  class Amp\Parallel\Context\ContextException#48 (8) {
    protected $message =>
    string(80) "The context stopped responding, potentially due to a fatal error or calling exit"
    private $string =>
    string(0) ""
    protected $code =>
    int(0)
    protected $file =>
    string(51) "C:\á\vendor\amphp\parallel\lib\Context\Process.php"
    protected $line =>
    int(177)
    private $trace =>
    array(16) {
      [0] =>
      array(4) {
        ...
      }
      [1] =>
      array(6) {
        ...
      }
      [2] =>
      array(6) {
        ...
      }
      [3] =>
      array(6) {
        ...
      }
      [4] =>
      array(6) {
        ...
      }
      [5] =>
      array(6) {
        ...
      }
      [6] =>
      array(6) {
        ...
      }
      [7] =>
      array(6) {
        ...
      }
      [8] =>
      array(6) {
        ...
      }
      [9] =>
      array(6) {
        ...
      }
      [10] =>
      array(6) {
        ...
      }
      [11] =>
      array(6) {
        ...
      }
      [12] =>
      array(6) {
        ...
      }
      [13] =>
      array(6) {
        ...
      }
      [14] =>
      array(6) {
        ...
      }
      [15] =>
      array(4) {
        ...
      }
    }
    private $previous =>
    class Amp\Parallel\Sync\ChannelException#28 (8) {
      protected $message =>
      string(136) "Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa"
      private $string =>
      string(0) ""
      protected $code =>
      int(0)
      protected $file =>
      string(54) "C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php"
      protected $line =>
      int(61)
      private $trace =>
      array(15) {
        ...
      }
      private $previous =>
      NULL
      public $xdebug_message =>
      string(1943) "
Amp\Parallel\Sync\ChannelException: Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa in C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php on line 61

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2"...
    }
    public $xdebug_message =>
    string(2297) "
Amp\Parallel\Context\ContextException: The context stopped responding, potentially due to a fatal error or calling exit in C:\á\vendor\amphp\parallel\lib\Context\Process.php on line 177

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2456    1371824   4. Amp\Loop\NativeDriver->run() C:\á\"...
  }
  [1] =>
  class Amp\Parallel\Context\ContextException#111 (8) {
    protected $message =>
    string(80) "The context stopped responding, potentially due to a fatal error or calling exit"
    private $string =>
    string(0) ""
    protected $code =>
    int(0)
    protected $file =>
    string(51) "C:\á\vendor\amphp\parallel\lib\Context\Process.php"
    protected $line =>
    int(177)
    private $trace =>
    array(16) {
      [0] =>
      array(4) {
        ...
      }
      [1] =>
      array(6) {
        ...
      }
      [2] =>
      array(6) {
        ...
      }
      [3] =>
      array(6) {
        ...
      }
      [4] =>
      array(6) {
        ...
      }
      [5] =>
      array(6) {
        ...
      }
      [6] =>
      array(6) {
        ...
      }
      [7] =>
      array(6) {
        ...
      }
      [8] =>
      array(6) {
        ...
      }
      [9] =>
      array(6) {
        ...
      }
      [10] =>
      array(6) {
        ...
      }
      [11] =>
      array(6) {
        ...
      }
      [12] =>
      array(6) {
        ...
      }
      [13] =>
      array(6) {
        ...
      }
      [14] =>
      array(6) {
        ...
      }
      [15] =>
      array(4) {
        ...
      }
    }
    private $previous =>
    class Amp\Parallel\Sync\ChannelException#220 (8) {
      protected $message =>
      string(136) "Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa"
      private $string =>
      string(0) ""
      protected $code =>
      int(0)
      protected $file =>
      string(54) "C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php"
      protected $line =>
      int(61)
      private $trace =>
      array(15) {
        ...
      }
      private $previous =>
      NULL
      public $xdebug_message =>
      string(1943) "
Amp\Parallel\Sync\ChannelException: Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa in C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php on line 61

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2"...
    }
    public $xdebug_message =>
    string(2297) "
Amp\Parallel\Context\ContextException: The context stopped responding, potentially due to a fatal error or calling exit in C:\á\vendor\amphp\parallel\lib\Context\Process.php on line 177

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2456    1371824   4. Amp\Loop\NativeDriver->run() C:\á\"...
  }
  [2] =>
  class Amp\Parallel\Context\ContextException#174 (8) {
    protected $message =>
    string(80) "The context stopped responding, potentially due to a fatal error or calling exit"
    private $string =>
    string(0) ""
    protected $code =>
    int(0)
    protected $file =>
    string(51) "C:\á\vendor\amphp\parallel\lib\Context\Process.php"
    protected $line =>
    int(177)
    private $trace =>
    array(16) {
      [0] =>
      array(4) {
        ...
      }
      [1] =>
      array(6) {
        ...
      }
      [2] =>
      array(6) {
        ...
      }
      [3] =>
      array(6) {
        ...
      }
      [4] =>
      array(6) {
        ...
      }
      [5] =>
      array(6) {
        ...
      }
      [6] =>
      array(6) {
        ...
      }
      [7] =>
      array(6) {
        ...
      }
      [8] =>
      array(6) {
        ...
      }
      [9] =>
      array(6) {
        ...
      }
      [10] =>
      array(6) {
        ...
      }
      [11] =>
      array(6) {
        ...
      }
      [12] =>
      array(6) {
        ...
      }
      [13] =>
      array(6) {
        ...
      }
      [14] =>
      array(6) {
        ...
      }
      [15] =>
      array(4) {
        ...
      }
    }
    private $previous =>
    class Amp\Parallel\Sync\ChannelException#92 (8) {
      protected $message =>
      string(136) "Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa"
      private $string =>
      string(0) ""
      protected $code =>
      int(0)
      protected $file =>
      string(54) "C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php"
      protected $line =>
      int(61)
      private $trace =>
      array(15) {
        ...
      }
      private $previous =>
      NULL
      public $xdebug_message =>
      string(1943) "
Amp\Parallel\Sync\ChannelException: Invalid packet received: Could not open input file: C:\\xc4\x82\xcb\x87\vendor\amphp\parallel\lib\Context/Internal/process-runner.php\xa in C:\á\vendor\amphp\parallel\lib\Sync\ChannelParser.php on line 61

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2"...
    }
    public $xdebug_message =>
    string(2297) "
Amp\Parallel\Context\ContextException: The context stopped responding, potentially due to a fatal error or calling exit in C:\á\vendor\amphp\parallel\lib\Context\Process.php on line 177

Call Stack:
    0.2030     381752   1. {main}() C:\á\examples\1-simple-function.php:0
    0.2456    1370816   2. Amp\Promise\wait() C:\á\examples\1-simple-function.php:15
    0.2456    1371384   3. Amp\Loop::run() C:\á\vendor\amphp\amp\lib\functions.php:158
    0.2456    1371824   4. Amp\Loop\NativeDriver->run() C:\á\"...
  }
}
```

**Comment by DaveRandom** [comment]
This is clearly an encoding issue, but I am unable to reconcile these dumps with the behaviour I see. I haven't run this through to it's conclusion yet, initially I'm just looking at how PHP handles it. In PHP 7.2 (x64 thread safe build from windows.php.net) I created a directory named `C:\á` and ran the following script in it:

```php
foreach (str_split(__DIR__) as $char) {
    printf(' %02X', ord($char));
}
```

This resulted, on my system (configured to English/UK) in the bytes below, I've added notation to show what they correspond to:

```
43 3A 5C C3 A1
C : \   á
```

This is UTF-8 encoded, in NF(K)C. However, the error messages above do not seem to indicate the same encoding, in fact they seem to be somewhat mangled - I cannot find any set of encoding/conversions that would result in the byte sequence `\xc4\x82\xcb\x87`.

@ostrolucky please could you run the above script on your system and confirm whether you get the same/different output?

**Comment by kelunik** [comment]
@DaveRandom According to http://www.utf8-chartable.de/unicode-utf8-table.pl?start=256&number=512&names=-&utf8=string-literal, it could to be a combination of `Ă`  (`\xc4\x82`) and `ˇ` (`\xcb\x87`), but that isn't the correct character.

**Comment by DaveRandom** [comment]
I suspect that somewhere along the line, `\xC3\xA1` has been interpreted as two characters and converted into some ANSI codepage (although I cannot for the life of me fathom what) and then possibly been re-interpreted as something else. The data is passed around quite a bit in various ways without any real thought being given to encoding, so really it's just a case of ensuring that everything agrees on what sort of data it expects.

However, I don't really want to put any more effort into it until I can confirm that PHP is at least seeing the same representations in a reasonably portable way, otherwise this is going to require a deep-dive into a lot of places.

**Comment by ostrolucky** [comment]
> @ostrolucky please could you run the above script on your system and confirm whether you get the same/different output?

Nope, sorry. I was at my parents during weekend. I don't have access to windows pc normally and won't have one until I come back, which will be maybe in 2 months from now. But I have written reproducer in my initial post, are you not able to reproduce it on windows pc?

**Comment by DaveRandom** [comment]
@ostrolucky I was basically trying to avoid having to do proper research but I guess I'll have to :-P

I'm going to consult with a few people who know more about PHP-windows internals than I do.

**Comment by ostrolucky** [comment]
You can also check reactphp codebase. I didn't have such issue there. But then, I am not sure it used such forking technique like amp/parallels used. I didn't use forking stuff directly, it was used by amp behind the scenes when I was working with sockets. I actually occurred this problem originally when script was crashing in moment client connected. Similar code worked fine with reactphp/http. If react doesn't use forking there, that should be another area where you can improve it. I feel amp libraries have too much dependencies between each other.

**Comment by ostrolucky** [comment]
Ok so reactphp/http doesn't trigger similar code because it doesn't use `hosts` file, like amphp/dns tries try to do. This is called via dependency amphp/socket -> amphp/dns -> amphp/file -> amphp/parallel. So that won't help you.

**Comment by kelunik** [comment]
@ostrolucky Right, ReactPHP does a blocking read for the hosts file.

**Comment by ostrolucky** [comment]
@DaveRandom So I'm back on that old PC just for this moment. I have run the script and I confirm I receive same output as you do

**Comment by kelunik** [comment]
https://bugs.php.net/bug.php?id=76333 might be related?

**Comment by ostrolucky** [comment]
Haha, I'm the one who reported that bug, from very same PC. So might be

**Comment by ostrolucky** [comment]
PHP 7.2.8, still a problem. IMO issue is in ProcessWrapper, not PHP. Was there more proper debugging done yet? At least running the example given in windows OS in "á" directory?

**Comment by ostrolucky** [comment]
So yeah, problem is in ProcessWrapper. 

`globals.exe_command_line` contains `"C:\tools\php72\php.exe" -dhtml_errors=0 -ddisplay_errors=0 -dlog_errors=1 "C:\Users\gadelat\Documents\á\stdinho\vendor\amphp\parallel\lib\Context/Internal/process-runner.php" "C:\Users\gadelat\Documents\á\stdinho\vendor\amphp\parallel\lib\Worker/Internal/worker-process.php" "Amp\Parallel\Worker\BasicEnvironment"`

but passing this as-is to [CreateProcess](https://github.com/amphp/windows-process-wrapper/blob/691e0cc6f1639d6ec4026d44aefa1738ccbf1031/src/ProcessWrapper.c#L427) does not work. Converting it to string and pasting it as such works.

```diff
+ "\"C:\\tools\\php72\\php.exe\" -dhtml_errors=0 -ddisplay_errors=0 -dlog_errors=1 \"C:\\Users\\gadelat\\Documents\\á\\stdinho\\vendor\\amphp\\parallel\\lib\\Context/Internal/process-runner.php\" \"C:\\Users\\gadelat\\Documents\\á\\stdinho\\vendor\\amphp\\parallel\\lib\\Worker/Internal/worker-process.php\" \"Amp\\Parallel\\Worker\\BasicEnvironment\"",
+ //globals.exe_command_line,
- globals.exe_command_line,
```

Some utf handling is needed. Unfortunately I'm no C programmer.

ping @DaveRandom @kelunik

**Comment by ostrolucky** [comment]
I created two php scripts, one using proc_open, second STDIN. First opens second script and passess UTF8 string. Works seemlessly. Will windows process wrapper be ever fixed, or it's totally unmaintained as it seems?

**Comment by ostrolucky** [comment]
I'm Linux person myself. I've got rid of separate Windows partition over 8 years ago. I've specifically created VM with Windows though just for this issue. It's really not an issue to reproduce this, even if you don't use Windows. This team created windows abstraction so it feels embarrassing to see you don't have resources to actually run it yourself.

It seems to me that this C project wasn't a good choice. It's simple project but more complicated to maintain, since it requires knowledge of low level language and you need to do lot of stuff yourself. It could be written in PHP instead - would result in having less LOC, getting rid of nasty binary, easier debugging and more people capable to maintain it.

**Comment by kelunik** [comment]
@ostrolucky We have that binary exactly, because it's not possible to write it in PHP. The STDIO pipes are blocking and we need real parallelism to read from all of them concurrently without blocking on a single one of them.

**Comment by ostrolucky** [comment]
What I understood that is being done is that PHP forks execution of the C binary and communicates with it. What could instead happen is PHP forks execution of PHP binary (PHP script).

**Comment by techno-express** [comment]
this binary is unnecessary take an look at http://php.net/manual/en/function.proc-open.php#120148 and https://bugs.php.net/bug.php?id=65650.

I was going to use this library found if this module is being used, i find it not safe, if it's not an regular php extension. It seems it will mean an whole rewrite to be usable. I am regular windows user, ssh into Linux dev boxes. 

I will need to see php script that hangs, that is not doing so because the way it's written.

**Comment by trowski** [comment]
@ostrolucky Using `pcntl_fork()` to create workers is not a good idea. It brings along the baggage of copying the entirety of the current environment, including any globals, references to other processes, etc. Additionally, `pcntl_fork()` is not available on Windows. I think you may not understand the fundamental underlying problem: process pipes are *always* blocking on Windows. Due to this problem, a wrapper is used on Windows that is able to poll STDOUT and STDERR simultaneously, communicating with the parent PHP process via sockets that can be non-blocking.

@techno-express Can you please clarify what you're talking about? The documentation comment and bug report you linked to have almost nothing to do with this issue. If you are having a specific issue, please open a new issue so we can address it separately.

**Comment by kelunik** [comment]
@techno-express The code you linked does a busy wait on the running processes and only reads the output once the process exits. There are no concurrent reads on multiple process outputs, which is the problem the wrapper solves on Windows.

**Comment by kelunik** [comment]
@ostrolucky: The PHP wrapper can still not read STDOUT and STDERR of its child concurrently, we need real threads there.

**Comment by kelunik** [comment]
@techno-express We don't owe you an example. If you want to improve this library, how about that: You write an example that reads concurrently from a set of child processes' STDOUT and STDERR with `stream_select`? I can probably invest a little time then to have a look and point out the issue (if it still exists with your solution).

**Comment by techno-express** [comment]
Ok, I will. It will show up here as mentions. I been working on another library locally on Windows. I follow everything related to async php. Make comments as points of reference.

To be clear, i never expected you to give an example. You haven't pointed the source of where this hanging issue has be raised, that you solution solves. It needs to be documented.

**Comment by kelunik** [comment]
Before we document these minor implementation details, we need to document the library as a whole.

**Comment by kelunik** [comment]
This has already been fixed via https://github.com/amphp/process/pull/34.

**Comment by ostrolucky** [comment]
correct
